test/test_Quaternion.jl

```diff
@@ -67,22 +67,79 @@ let # test rotations
-        unary_funs = [exp, log, sin, cos, sqrt, inv, conj, abs2, norm]
-        # since every quaternion is conjugate to a complex number,
-        # one can establish correctness as follows:
-        for fun in unary_funs
-            @test fun(Quaternion(c)) ≈ Quaternion(fun(c))
+@testset "conversions" begin
+    z = rand(ComplexF64)
+    q = convert(Quaternion{Float64}, z)
+    @test q.s == real(z)
+    @test q.v1 == imag(z)
+    @test q.v2 == 0
+    @test q.v3 == 0
+end
+
+@testset "non-analytic functions" begin
+    q, q2 = randn(Quaternion{Float64}, 2)
+    unary_funs = [conj, abs, abs2, norm, sign]
+    # since every quaternion is conjugate to a complex number,
+    # one can establish correctness as follows:
+    @testset for fun in unary_funs
+        for i in 1:100
+            c = randn(ComplexF64)
+            @test fun(Quaternion(c)) ≈ fun(c)
+            @test q2 * fun(q) * inv(q2) ≈ fun(q2 * q * inv(q2))
+        end
+    end
+end
+
+@testset "complex analytic functions" begin
+    # all complex analytic functions can be extended to the quaternions
+    q, q2 = randn(Quaternion{Float64}, 2)
+    unary_funs = [
+        sqrt, inv, exp, exp2, exp10, expm1, log, log2, log10, log1p,
+        sin, cos, tan, asin, acos, atan, sinh, cosh, tanh, asinh, acosh, atanh,
+        csc, sec, cot, acsc, asec, acot, csch, sech, coth, acsch, asech, acoth,
+        sinpi, cospi,
+    ]
+    # since every quaternion is conjugate to a complex number,
+    # one can establish correctness as follows:
+    @testset for fun in unary_funs
+        for i in 1:100
+            c = randn(ComplexF64)
+            @test fun(Quaternion(c)) ≈ fun(c)
             @test q2 * fun(q) * inv(q2) ≈ fun(q2 * q * inv(q2))
         end
+    end
 
-        @test exp(log(q)) ≈ q
-        @test exp(zero(q)) ≈ one(q)
+    q = randn(Quaternion{Float64})
+    @test inv(q) * q ≈ q * inv(q) ≈ one(q)
+    @test sqrt(q) * sqrt(q) ≈ q
+    @test exp(log(q)) ≈ q
+    @test exp(zero(q)) ≈ one(q)
+    @test log(one(q)) ≈ zero(q)
+    @test exp2(log2(q)) ≈ q
+    @test exp10(log10(q)) ≈ q
+    @test expm1(log1p(q)) ≈ q
+    @test all(sincos(q) .≈ (sin(q), cos(q)))
+    VERSION ≥ v"1.6" && @test all(sincospi(q) .≈ (sinpi(q), cospi(q)))
+    @test tan(q) ≈ cos(q) \ sin(q) ≈ sin(q) / cos(q)
+    @test tanh(q) ≈ cosh(q) \ sinh(q) ≈ sinh(q) / cosh(q)
+    @testset for (f, finv) in [(sin, csc), (cos, sec), (tan, cot), (sinh, csch), (cosh, sech), (tanh, coth)]
+        @test f(q) ≈ inv(finv(q))
     end
+    @testset for (f, finv) in [(asin, acsc), (acos, asec), (atan, acot), (asinh, acsch), (acosh, asech), (atanh, acoth)]
+        @test f(q) ≈ finv(inv(q))
+    end
+    @test q^1.3 ≈ exp(1.3 * log(q))
+    @test q^7.8 ≈ exp(7.8 * log(q))
+    @test q^1.3f0 ≈ exp(1.3f0 * log(q))
+    @test q^7.8f0 ≈ exp(7.8f0 * log(q))
+end
 
+let
+    # https://github.com/JuliaGeometry/Quaternions.jl/issues/39
+    @test exp(Quaternion(1,1,1,1)) ≈ exp(Quaternion(1.0,1.0,1.0,1.0))
+end
```

Comment:
```suggestion
@testset "exp(::Quaternion{Int})" begin
    # https://github.com/JuliaGeometry/Quaternions.jl/issues/39
    @test exp(Quaternion(1,1,1,1)) ≈ exp(Quaternion(1.0,1.0,1.0,1.0))
end
```
Do we need this `let` block here? I thought it's better to remove `let` or replace it with `@testset`.